internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go

```diff
@@ -21,91 +23,389 @@ import (
 // BuildDestinations creates listeners, routers, clusters, and endpointRefs for all destinations
 // and adds them to the proxyState.
 func (b *Builder) BuildDestinations(destinations []*intermediate.Destination) *Builder {
+	var lb *ListenerBuilder
 	if b.proxyCfg.GetDynamicConfig() != nil &&
 		b.proxyCfg.DynamicConfig.Mode == pbmesh.ProxyMode_PROXY_MODE_TRANSPARENT {
-		b.addTransparentProxyOutboundListener(b.proxyCfg.DynamicConfig.TransparentProxy.OutboundListenerPort)
+		lb = b.addTransparentProxyOutboundListener(b.proxyCfg.DynamicConfig.TransparentProxy.OutboundListenerPort)
 	}
 
 	for _, destination := range destinations {
+		b.buildDestination(lb, destination)
+	}
+
+	return b
+}
+
+func (b *Builder) buildDestination(
+	tproxyOutboundListenerBuilder *ListenerBuilder,
+	destination *intermediate.Destination,
+) *Builder {
+	var (
+		// Find the destination proxy's port.
+		//
+		// Endpoints refs will need to route to mesh port instead of the
+		// destination port as that is the port of the destination's proxy.
+		meshPortName      string
+		virtualPortNumber uint32
+		effectiveProtocol = destination.ComputedPortRoutes.Protocol
+		targets           = destination.ComputedPortRoutes.Targets
```

Comment:
Each key in the computed routes targets map will turn into a Cluster (or after failover works, an aggregate cluster).